src/smc-webapp/project_files.cjsx

```diff
@@ -1825,11 +1829,70 @@ ProjectFiles = (name) -> rclass
                 style   = {error_style}
                 onClose = {=>@props.actions.setState(error:'')} />
 
+    handle_keypress : (e) ->
+        if e.keyCode == 13
+            @setState(reason_for_requesting_to_join_project:@refs.reason_for_requesting_to_join.getValue())
+            @send_request_to_join_project()
+
+    send_request_to_join_project : ->
+        @setState(is_request_to_join_project_modal_open:false)
+        if smc.redux.getStore('projects').get_project(@props.project_id)?.invite_requests
+            invite_requests = smc.redux.getStore('projects').get_project(@props.project_id).invite_requests
+        else
+            invite_requests = {}
+        invite_requests[smc.client.account_id] = {timestamp:salvus_client.server_time(), message:@state['reason_for_requesting_to_join_project']}
+        smc.client.query({cb:console.log, query:{project_invite_requests:{project_id:@props.project_id, invite_requests:invite_requests}}})
+
+    request_to_join_project : ->
+        @setState(is_request_to_join_project_modal_open:true)
+
+    render_request_to_join_project_modal : ->
+        <div className="request_to_join_project_modal">
+            <Modal show={@state['is_request_to_join_project_modal_open']} onHide={=>@setState(is_request_to_join_project_modal_open:false)}>
+                <Modal.Header>
+                    <Modal.Title>Request to join project</Modal.Title>
+                </Modal.Header>
+
+              <Modal.Body>
+                <Input
+                    ref         = 'reason_for_requesting_to_join'
+                    type        = 'text'
+                    placeholder = {'Reason for requesting to join project'}
+                    onChange    = {=>@setState(reason_for_requesting_to_join_project:@refs.reason_for_requesting_to_join.getValue())}
+                    onKeyDown   = {@handle_keypress}
+                />
+              </Modal.Body>
+
+              <Modal.Footer>
+                <Button onClick={=>@setState(is_request_to_join_project_modal_open:false)}>Close</Button>
+                <Button bsStyle="primary" onClick={=>@send_request_to_join_project()}>Send request to join project</Button>
+              </Modal.Footer>
+
+            </Modal>
+        </div>
+
+    render_request_to_join_project_button : ->
+        <Button onClick={=>@request_to_join_project()}>Request to join this project</Button>
+
+    render_not_public_error : ->
+        if @props.redux.getStore('account').is_logged_in()
+            <div>
+                <ErrorDisplay title="Directory is not public" error={"You are trying to access a non public project that you are not a collaborator on. You need to ask a collaborator of the project to add you."} />
+                {if not smc.client.account_id in smc.redux.getStore('projects').get_project(@props.project_id)?.invite_requests then @render_request_to_join_project_button()}
```

Comment:
This line is also wrong.  It'll sort of work if you have been testing for a while, but for a clean test it will fail since invite_requests is undefined.  I'll fix this.    In the future, be more careful in testing.